starlark/int.go

```diff
@@ -8,74 +8,77 @@ import (
 	"math"
 	"math/big"
+	"strconv"
 
 	"go.starlark.net/syntax"
 )
 
 // Int is the type of a Starlark int.
-type Int struct{ bigint *big.Int }
+type Int struct {
+	// big != nil <=> value is not representable as int32.
+	// We use only the signed 32 bit range of small to ensure
+	// that small+small and small*small do not overflow.
+
+	small int64    // minint32 <= small <= maxint32
+	big   *big.Int // big.BitLen() > 32
+}
+
+// newBig allocates a new big.Int.
+func newBig(x int64) *big.Int {
+	if 0 <= x && int64(big.Word(x)) == x {
+		// x is guaranteed to fit into a single big.Word.
+		// Most starlark ints are small,
+		// but math/big assumes that since you've chosen to use math/big,
+		// your big.Ints will probably grow, so it over-allocates.
+		// Avoid that over-allocation by manually constructing a single-word slice.
+		// See https://golang.org/cl/150999, which will hopefully land in Go 1.13.
+		return new(big.Int).SetBits([]big.Word{big.Word(x)})
+	}
+	return big.NewInt(x)
+}
 
 // MakeInt returns a Starlark int for the specified signed integer.
 func MakeInt(x int) Int { return MakeInt64(int64(x)) }
 
 // MakeInt64 returns a Starlark int for the specified int64.
 func MakeInt64(x int64) Int {
-	if 0 <= x {
-		if x < int64(len(smallint)) {
-			if !smallintok {
-				panic("MakeInt64 used before initialization")
-			}
-			return Int{&smallint[x]}
-		}
-		if int64(big.Word(x)) == x {
-			// x is guaranteed to fit into a single big.Word.
-			// Most starlark ints are small,
-			// but math/big assumes that since you've chosen to use math/big,
-			// your big.Ints will probably grow, so it over-allocates.
-			// Avoid that over-allocation by manually constructing a single-word slice.
-			// See https://golang.org/cl/150999, which will hopefully land in Go 1.13.
-			return Int{new(big.Int).SetBits([]big.Word{big.Word(x)})}
-		}
+	if math.MinInt32 <= x && x <= math.MaxInt32 {
+		return Int{small: x}
 	}
-	return Int{big.NewInt(x)}
+	return Int{big: newBig(x)}
 }
 
 // MakeUint returns a Starlark int for the specified unsigned integer.
 func MakeUint(x uint) Int { return MakeUint64(uint64(x)) }
 
 // MakeUint64 returns a Starlark int for the specified uint64.
 func MakeUint64(x uint64) Int {
-	if x < uint64(len(smallint)) {
-		if !smallintok {
-			panic("MakeUint64 used before initialization")
-		}
-		return Int{&smallint[x]}
+	if x <= math.MaxInt32 {
+		return Int{small: int64(x)}
 	}
 	if uint64(big.Word(x)) == x {
-		// See comment in MakeInt64 for an explanation of this optimization.
-		return Int{new(big.Int).SetBits([]big.Word{big.Word(x)})}
+		// See comment in newBig for an explanation of this optimization.
+		return Int{big: new(big.Int).SetBits([]big.Word{big.Word(x)})}
 	}
-	return Int{new(big.Int).SetUint64(x)}
+	return Int{big: new(big.Int).SetUint64(x)}
+}
+
+// MakeBigInt returns a Starlark int for the specified big.Int.
+// The caller must not subsequently modify x.
+func MakeBigInt(x *big.Int) Int {
+	if x.BitLen() < 32 {
```

Comment:
This will have to be updated to:

n := x.BitLen()
if n < 32 || n == 32 && x.Int64() == math.MinInt32 {